primary/src/helper.rs

```diff
@@ -63,35 +72,128 @@ impl<PublicKey: VerifyingKey> Helper<PublicKey> {
                     self.process_certificates(certificate_ids, requestor, true)
                         .await;
                 }
+                // A request that another primary sends us to check whether we have
+                // available to serve the payloads for the provided certificate_ids.
+                PrimaryMessage::PayloadAvailabilityRequest {
+                    certificate_ids,
+                    requestor,
+                } => {
+                    self.process_payload_availability(certificate_ids, requestor)
+                        .await;
+                }
                 _ => {
                     panic!("Received unexpected message!");
                 }
             }
         }
     }
 
+    /// Checks for each provided digest of the list of digests they node has
+    /// the corresponding certificate within its storage and the payload as
+    /// well. If yes, then the requestor can follow up and request the batches
+    /// payload from the primary's workers via the standard syncing approach.
+    async fn process_payload_availability(
+        &mut self,
+        digests: Vec<CertificateDigest>,
+        origin: PublicKey,
+    ) {
+        // get the requestor's address.
+        let address = match self.committee.primary(&origin) {
+            Ok(x) => x.primary_to_primary,
+            Err(e) => {
+                warn!("Primary origin node not found in committee: {e}");
+                return;
+            }
+        };
+
+        let mut result: Vec<(CertificateDigest, bool)> = Vec::new();
+
+        let certificates = match self.certificate_store.read_all(digests.to_owned()).await {
+            Ok(certificates) => certificates,
+            Err(err) => {
+                error!("Error while retrieving certificates: {err}");
```

Comment:
This is a failure on a simple DB read: we're going to have many such ones in logs. 
We can enter a [span](https://docs.rs/tracing/latest/tracing/#spans), equipped with context from the digests, to help us log what happened here in a precise way without encumbering this function with logging code. 

Here's an example of it in action in Sui:
https://github.com/MystenLabs/sui/blob/1a25e68fbc849684344680a4da134d6a62b136c3/sui_core/src/authority_server.rs#L169-L174

Reply:
That a really good idea. Ok let me address this to a follow up PR so I can deal with tracing a bit more properly. I'll actually going to look into how I could do this in a more centralised way - meaning starting the tracing from the request level - will take the `helper` as an example and provide something.

Reply:
Issue here: https://github.com/MystenLabs/narwhal/issues/201